docs/jit.md

```diff
@@ -0,0 +1,255 @@
+* [2] http://openjdk.java.net/legal/assembly-exception.html
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH
+* Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# The JIT compiler
+
+The Just-In-Time (JIT) compiler is a key component of the OpenJ9 VM that improves the performance of Java applications by compiling platform-neutral Java bytecode into native machine code at run time. Without the JIT, the VM has to interpret the bytecodes itself - a process that requires extra CPU and memory.
+
+The JIT compiler doesn't compile every method that gets called because thousands of methods can be called at startup. Instead, OpenJ9 records the number of times a method is called. When the count reaches a pre-defined *invocation threshold*, JIT compilation is triggered. Once a method has been compiled by the JIT, the VM can call the compiled method rather than interpreting it.
+
+## Optimization levels
+
+The JIT compiler can compile a method at different optimization levels: **cold**, **warm**, **hot**, **very hot (with profiling)**, or **scorching**. The hotter the optimization level, the better the expected performance, but the higher the cost in terms of CPU and memory.
+
+- **cold** is used during startup processing for large applications where the goal is to achieve the best compiled code speed for as many methods as possible.
+- **warm** is the workhorse; after start-up, most methods are compiled when they reach the invocation threshold.
+
+For higher optimization levels, the VM uses a sampling thread to identify methods that continue to take a lot of time. Methods that consume more than 1% are compiled at hot. Methods that consume more than 12.5% are scheduled for a scorching compilation. However, before that happens the methods are compiled at very hot with profiling to collect detailed profile data that is used by the scorching compilation.
+
+The higher optimization levels use special techniques such as escape analysis and partial redundancy elimination, or loop through certain optimization sequences more times. Although these techniques use more CPU and memory, the improved performance that is delivered by the optimizations can make the tradeoff worthwhile.
+
+## Troubleshooting
+
+The JIT compiler is enabled by default to optimize performance. However, if you experience a problem running your application, temporarily turning off the JIT will tell you whether the JIT is at fault.
+
+Because JIT starts at the same time as the VM, you can only modify JIT behavior at startup.
+
+There are a number of ways to disable the JIT:
+
+- Specify `-Djava.compiler=NONE` on the command line.
+- Specify `-Xint` on the command line, which turns off the JIT and AOT compiler. To eliminate problems with one or the other you can turn these compilers off selectively with the `-Xnojit` and `-Xnoaot` options.
+- Call the `java.lang.Compiler` API programmatically.
```

Comment:
Note the java.lang.Compile class is deprecated for removal since Java 9.

Reply:
Fixed here.